### PR TITLE
Various code and documentation improvements

### PR DESCRIPTION
**Pull request by cweniger**
There are a range of changes

- Removed `Task` and `Microscope`, which apparently were not used and only cause confusion, since they hide some of the internal logic.
- Added *many* docstrings. Still not finished though. I follow the same style as pytorch.
- Fix Sphinx documentation compilation issues. Should work now, hopefully also for readthedocs.
- Updated autodocs, with better ordering of relevant functions, and exposure of inherited methods for MemoryStore etc.
- Clarified `params`: We used `params` and `pars` for all kinds of thinks in an inconsistent way. Now there are is is `pnames` for parameters names, `v` for the actual parameter vectors, `u` for the hypercube projection, `, `pois` for "parameters of interest"
- Model parameter names are now specified in `swyft.Simulator`, and consistently propagated into the `Store`, `Dataset` and `Posteriors`.
- We removed `train_args` and replaced It with `TrainOptions` internally. Defaults are uniquely specified as keyword arguments of `Posteriors.train` method, which makes them transparent to the user.
- `Posteriors` is now instantiated via a `Dataset` object, which is then also the (overwritable) default for training. This makes the `Task` object obsolete (which was only a thin layer around Posteriors that led to lots of code duplication).
- Fixed many of the `load` and `save` functions.
- Streamlining of `MemoryStore` and `DirectoryStore`, which had some redundant functionality in terms of saving and loading.
- Separating `sample`, `coverage`, and `add` methods of `Store`.
- Parameter names are now stored inside the `Store`. Before they had to be re-specified each time when loading the store.
- Made some of the store access functions faster.
- Samples now also return `u` and parameter names, which is useful for plotting routines.

### CODE REVIEW

**Comment by cweniger** [comment]
Still to do:
- [ ] Document limitations of Prior mapping (add warning).
- [ ] DEVICE in `Posteriors.add` or `Posteriors.train`?
- [ ] Write "marginals" in quickstart.

File: swyft/store/store.py [hunk 1]
```diff
@@ -420,82 +463,58 @@ def wait_for_simulations(self, indices):
             done = np.all(done)
 
 
-#    # FIXME: Necessary
```
Comment (bkmi) [inline swyft/store/store.py:423]: [ ] Combine the DirectoryStore and MemoryStore into one class with factory methods

File: swyft/store/dataset.py [hunk 2]
```diff
@@ -10,104 +10,125 @@
...
+        """Return prior of dataset (swyft.Prior)."""
         return self._trunc_prior.prior
 
     @property
     def bound(self):
+        """Return bound of truncated prior of dataset (swyft.Bound)."""
         return self._trunc_prior.bound
 
     def _tensorfy(self, x):
         return {k: torch.tensor(v).float() for k, v in x.items()}
```
Comment (bkmi) [inline swyft/store/dataset.py:63]: [ ] we should use the function `array_to_tensor` whenever we convert from x to tensor.

File: swyft/inference/posteriors.py [hunk 3]
```diff
@@ -47,7 +49,7 @@ def sample(self, N, obs, device=None, n_batch=10_000):
 
             weights[k] = np.exp(val)
 
-        return dict(params=v, weights=weights)
+        return dict(v=v, weights=weights)
 
     def rejection_sample(
```
Comment (bkmi) [inline swyft/inference/posteriors.py:54]: [ ]This should be moved over to Posteriors rather than living in PosteriorCollection

Comment (bkmi) [inline swyft/inference/posteriors.py:54]: [] This should be moved over to Posteriors rather than living in PosteriorCollection

**Comment by bkmi** [comment]
[ ] add type annotations (add a comment that's searchable for @cweniger 
[ ] remove anything that says deprecated!

File: swyft/inference/posteriors.py (hunk 3 above)
Comment (bkmi) [reply swyft/inference/posteriors.py:54]: this meaning rejection_sample

File: swyft/store/store.py (hunk 1 above)
Comment (bkmi) [reply swyft/store/store.py:423]: We're not going to do this

**Comment by bkmi** [comment]
Shall we also remove the `scripts/` folder?

**Pull request merged**